gix-date/src/parse.rs

```diff
@@ -83,52 +83,79 @@ pub(crate) mod function {
         };
         Some(time)
     }
+
+    /// This is just like `Zoned::strptime`, but it allows parsing datetimes
+    /// whose weekdays are inconsistent with the date. While the day-of-week
+    /// still must be parsed, it is otherwise ignored. This seems to be
+    /// consistent with how `git` behaves.
+    fn strptime_relaxed(fmt: &str, input: &str) -> Result<Zoned, jiff::Error> {
+        let mut tm = jiff::fmt::strtime::parse(fmt, input)?;
+        tm.set_weekday(None);
+        tm.to_zoned()
+    }
+
+    /// This is just like strptime_relaxed, except for RFC 2822 parsing.
+    /// Namely, it permits the weekday to be inconsistent with the date.
+    fn rfc2822_relaxed(input: &str) -> Result<Zoned, jiff::Error> {
+        static P: rfc2822::DateTimeParser = rfc2822::DateTimeParser::new().relaxed_weekday(true);
+        P.parse_zoned(input)
+    }
 }
 
 mod relative {
     use std::{str::FromStr, time::SystemTime};
 
-    use time::{Duration, OffsetDateTime};
+    use jiff::{tz::TimeZone, Span, Timestamp, Zoned};
 
     use crate::parse::Error;
 
-    fn parse_inner(input: &str) -> Option<Duration> {
+    fn parse_inner(input: &str) -> Option<Result<Span, Error>> {
         let mut split = input.split_whitespace();
-        let multiplier = i64::from_str(split.next()?).ok()?;
+        let units = i64::from_str(split.next()?).ok()?;
         let period = split.next()?;
         if split.next()? != "ago" {
             return None;
         }
-        duration(period, multiplier)
+        span(period, units)
     }
 
-    pub(crate) fn parse(input: &str, now: Option<SystemTime>) -> Option<Result<OffsetDateTime, Error>> {
-        parse_inner(input).map(|offset| {
-            let offset = std::time::Duration::from_secs(offset.whole_seconds().try_into()?);
+    pub(crate) fn parse(input: &str, now: Option<SystemTime>) -> Option<Result<Zoned, Error>> {
+        parse_inner(input).map(|result| {
+            let span = result?;
+            // This was an error case in a previous version of this code, where
+            // it would fail when converting from a negative signed integer
+            // to an unsigned integer. This preserves that failure case even
+            // though the code below handles it okay.
```

Comment:
Yeah I noticed that because I was intensely curious as to why there was a `catch_unwind` here deep inside datetime parsing. And yeah indeed, I believe [`time`'s panic comes from its `From<SystemTime>` impl](https://github.com/time-rs/time/blob/aca5c3652aff6fc7df3fbd8591c3e324d3a6ffb2/time/src/offset_date_time.rs#L1565-L1566), which in turn calls [`OffsetDateTime`'s `Add` impl that specifically panics](https://github.com/time-rs/time/blob/aca5c3652aff6fc7df3fbd8591c3e324d3a6ffb2/time/src/offset_date_time.rs#L1362-L1368). The `Add` impl panicking makes sense, since that's consistent with what `std` does too. But the `From` impl should probably be a `TryFrom` impl. Since it's a `From` impl, `time`'s behavior here is pretty much locked into its API. With that said, I think you probably could have avoided the `catch_unwind` here by first converting `now` to a `time::OffsetDateTime` and then doing your arithmetic, since `time` exposes checked operations on `time::OffsetDateTime`. With Jiff, you can't mess this up because it doesn't provide a `From<SystemTime>` impl. Only `TryFrom<SystemTime>`.